denv.8
.\" Man page generated from reStructuredText.
.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.TH "DENV" "8" "" ""
.SH NAME
denv \- Run a command in a modified environment
.SH SYNOPSIS
.sp
\fBdenv [options] command [command\-options]\fP
.sp
\fBenvdir d child\fP
.sp
The second form mimics the command line interface of the \fIenvdir(8)\fP
tool included in DJB\(aqs daemontools package, see CAVEATS for the
differences.
.SH DESCRIPTION
.sp
The \fBdlog\fP program modifies the environment variables as specified by
the supplied \fBoptions\fP and immediately executes the given \fBcommand\fP\&.
.sp
By default the environment will be empty. Variables from the parent process
may be incorporated using the \fB\-\-inherit\-env\fP and \fB\-\-inherit\fP options,
and new variables added or removed using the \fB\-\-environ\fP, \fB\-\-direnv\fP
and \fB\-\-file\fP options.
.SH USAGE
.sp
Command line options:
.INDENT 0.0
.TP
.B  \-I\fP,\fB  \-\-inherit\-env
Inherit all the environment variables from the parent
process that runs \fBdenv\fP\&.
.TP
.BI \-i \ NAME\fR,\fB \ \-\-inherit \ NAME
Inherit and environment variable from the parent process
that runs \fBdenv\fP given its \fBNAME\fP\&. This option may
be used multiple times.
.UNINDENT
.INDENT 0.0
.TP
.B \-E NAME[=VALUE], \-\-environ NAME[=VALUE]
Define an environment variable with a given \fBNAME\fP, and
assign a \fBVALUE\fP to it. If the value is not specified,
an existing variable is removed.
.UNINDENT
.INDENT 0.0
.TP
.BI \-d \ PATH\fR,\fB \ \-\-envdir \ PATH
Read each file from the directory at \fBPATH\fP, assigning
as variables named after the files, with their values
being the first line of contents. Trailing spaces are
removed from the values. If a file (or its first line)
is empty, the variable named after the file is removed.
Only regular files are taken into account, and in
particular symbolic links are \fInot\fP followed.
.TP
.BI \-f \ PATH\fR,\fB \ \-\-file \ PATH
Read lines from the file at \fBPATH\fP, each one containing
a \fBNAME=[VALUE]\fP environment variable assignment. Both
leading and trailing whitespace are removed from the line.
Each \fBNAME\fP must be a valid variable name. Empty lines
and lines beginning with the comment character \fB#\fP are
ignored. If a \fBVALUE\fP is omitted, then the variable is
removed from the environment. The file format is loosely
based on \fIenvironment.d(5)\fP\&.
.UNINDENT
.sp
The \fBdenv\fP program is often used in concert with \fIdmon(8)\fP to setup
the environment under which supervised programs run.
.SH EXIT CODES
.sp
\fBdenv\fP exits with code \fB111\fP if it has trouble preparing the environment
before executing the chained child process, or if the child process cannot
be executed. Otherwise its exit code is that of the child process.
.SH SEE ALSO
.sp
\fIdmon(8)\fP, \fIenvdir(8)\fP, \fIenvironment.d(5)\fP
.sp
 <http://cr.yp.to/daemontools.html> 
.SH CAVEATS
.sp
The original \fIenvdir(8)\fP program included in the daemontools suite converts
null characters (\fB\e0\fP) into new lines (\fB\en\fP). This behaviour is
deliberately left unimplemented in \fBdenv\fP, even when it runs in \fBenvdir\fP
mode.
.SH AUTHOR
Adrian Perez <aperez@igalia.com>
.\" Generated by docutils manpage writer.
.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) 
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [v0.6.0] - 2024-12-10
### Added
- New `denv` utility, inspired by daemontools' `envdir`, which may be
  used to manipulate the environment before executing a chained child
  process.

### Fixed
- Retry system calls that may return with `errno` set to `EINTR` when
  interrupted by signals to indicate they can be restarted from user space.
- Remove leftover, useless `$E` format specifier when reporting errors
  during daemonization.

### Changed
- The project is now built in C23 mode by default. In practice `-std=c2x`
  gets used to cover systems which may have slightly older compilers.

## [v0.5.1] - 2021-02-23
### Added
- The `dlog`, `dslog`, and `drlog` tools now log empty input lines. The
  `--skip-empty`/`-e` command line option has been added to disable
  logging empty lines.
- The `dmon` tool has gained a `--max-respawns`/`-m` command line option which
  can be used to specify how many times to respawn monitored processes
  before exiting. (Patch by Matt Schulte <<mschulte@wyze.com>>.)
- New `libsetunbuf.so` helper which can be used to disable buffering of
  the standard output stream on arbitrary programs via `LD_PRELOAD`.
  (Patch by Matt Schulte <<mschulte@wyze.com>>.)

### Fixed
- The exit status of monitored processes is now correctly propagated as
  exit code of `dmon` itself.
- The `dlog`, `dslog`, and `drlog` tools will no longer exit unexpectedly
  when they receive an empty input line. (Patch by Matt Schulte
  <<mschulte@wyze.com>>.)

## [v0.5.0] - 2020-08-27
### Added
- Support listing applets compiled into a multicall `dmon` binary when the
  `DMON_LIST_MULTICALL_APPLETS` environment variable is set and non-zero.

### Changed
- GNU Make is not required anymore, the included `Makefile` now works with
  the BSD variant as well, and probably others.

## [v0.4.5] - 2018-08-05
### Fixed
- Make it possible to build `dmon` with newer GCC versions.

## [v0.4.4] - 2016-10-30
### Fixed
- Correctly forward signals to the log process when using `-S`/`--log-sigs`.

## [v0.4.3] - 2016-10-13
### Added
- Allow setting the work directory with `-W`/`--work-dir` in the command line
  (or `work-dir` in a configuration file).

## [v0.4.2] - 2012-01-07
### Added
- Manual pages now also include the long command line options.
- `dlog` can now be instructed to prefix every log message with a given string,
  using the `-p`/`--prefix` command line option.
- `dlog`, `drlog`, and `dslog` have gained the `-i`/`--input-fd` command line
  option, which allows reading log messages from file descriptors different from
  the standard input.

### Fixed
- `dlog` and `drlog` now handle the `INT`, `TERM`, and `HUP` signals gracefully.

[Unreleased]: https://github.com/aperezdc/dmon/compare/v0.6.0...HEAD
[v0.6.0]: https://github.com/aperezdc/dmon/compare/v0.5.1...v0.6.0
[v0.5.1]: https://github.com/aperezdc/dmon/compare/v0.5.0...v0.5.1
[v0.5.0]: https://github.com/aperezdc/dmon/compare/v0.4.5...v0.5.0
[v0.4.5]: https://github.com/aperezdc/dmon/compare/v0.4.4...v0.4.5
[v0.4.4]: https://github.com/aperezdc/dmon/compare/v0.4.3...v0.4.4
[v0.4.3]: https://github.com/aperezdc/dmon/compare/v0.4.2...v0.4.3
[v0.4.2]: https://github.com/aperezdc/dmon/compare/v0.4.1...v0.4.2


denv.c
/*
 * Copyright (C) 2024 Adrian Perez de Castro <aperez@igalia.com>
 *
 * SPDX-License-Identifier: MIT
 */

#ifndef __linux
#define _BSD_SOURCE
#endif

#define _DEFAULT_SOURCE
#define _GNU_SOURCE
#define _POSIX_C_SOURCE 200809L

#include "deps/cflag/cflag.h"
#include "deps/clog/clog.h"
#include "deps/dbuf/dbuf.h"
#include "util.h"

#include <assert.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

#if !(defined(MULTICALL) && MULTICALL)
# define denv_main main
#endif /* MULTICALL */

extern char **environ;

static char *argv0 = "denv";
static char **env = NULL;
static size_t env_a = 0;
static size_t env_n = 0;

static void
env_del(const char* name, size_t namelen)
{
	if (!env)
		return;

	size_t i;
	for (i = 0; env[i]; i++) {
		const char *equalsign = strchr(env[i], '=');
		assert(equalsign);

		size_t envlen = equalsign - env[i];
		if (envlen != namelen)
			continue;

		if (memcmp(env[i], name, namelen) == 0) {
			for (; env[i]; i++) {
				env[i] = env[i + 1];
			}
			assert(env_n > 0);
			env_n--;
			return;
		}
	}
}

static void
env_add(char *entry)
{
	assert(entry);

	const char *equalsign = strchr(entry, '=');
	assert(equalsign);

	if (env_n + 1 >= env_a) {
		env_a = env_a * 2 + 5;
		if (env) {
			env = reallocarray(env, env_a, sizeof(char*));
		} else {
			env = calloc(env_a, sizeof(char*));
		}
		if (!env)
			fatal("%s: Cannot allocate memory.\n", argv0);
	}
	assert(env_a > (env_n + 1));

	env_del(entry, equalsign - entry);
	env[env_n++] = entry;
	env[env_n] = NULL;
}

static void
env_inherit_all(void)
{
	for (size_t i = 0; environ[i]; i++) {
		assert(strchr(environ[i], '='));
		env_add(environ[i]);
	}
}

static enum cflag_status
opt_inherit_env(const struct cflag *spec, const char *arg)
{
	(void) arg;
	if (!spec)
		return CFLAG_OK;

	env_inherit_all();
	return CFLAG_OK;
}

static enum cflag_status
opt_inherit(const struct cflag *spec, const char *arg)
{
	if (!spec)
		return CFLAG_NEEDS_ARG;

	for (size_t i = 0; environ[i]; i++) {
		const char *equalsign = strchr(environ[i], '=');
		assert(equalsign);
		if (strncmp(arg, environ[i], equalsign - environ[i]) == 0) {
			env_add(environ[i]);
			return CFLAG_OK;
		}
	}

	clog_debug("Cannot inherit undefined variable '%s'", arg);
	return CFLAG_OK;
}

static enum cflag_status
opt_environ(const struct cflag *spec, const char *arg)
{
    if (!spec)
        return CFLAG_NEEDS_ARG;

    const char *equalsign;
    if ((equalsign = strchr(arg, '=')) == NULL) {
        env_del(arg, strlen(arg));
    } else {
		env_add((char*) arg);
    }
    return CFLAG_OK;
}

#ifndef AT_NO_AUTOMOUNT
#define AT_NO_AUTOMOUNT 0
#endif /* !AT_NO_AUTOMOUNT */

static inline bool
is_trim_char(int c)
{
	switch (c) {
		case '\r':
		case '\n':
		case '\v':
		case '\t':
		case '\f':
		case ' ':
			return true;
		default:
			return false;
	}
}

static void
env_envdir(const char *path)
{
	assert(path);

	DIR *d = opendir(path);
	if (!d)
		fatal("%s: Cannot open directory '%s': %s.\n", argv0, path, ERRSTR);

	struct dirent *de;
	while ((de = readdir(d))) {
		if (de->d_name[0] == '.')
			continue;

		struct stat st;
		if (fstatat(dirfd(d), de->d_name, &st, AT_NO_AUTOMOUNT | AT_SYMLINK_NOFOLLOW) == -1)
			fatal("%s: Cannot stat '%s/%s': %s.", argv0, path, de->d_name, ERRSTR);

		if (!S_ISREG(st.st_mode))
			continue;

		/* Remove from the environment if the file is empty. */
		if (st.st_size == 0) {
			env_del(de->d_name, strlen(de->d_name));
			continue;
		}

		int fd = safe_openat(dirfd(d), de->d_name, O_RDONLY);
		if (fd < 0)
			fatal("%s: Cannot open '%s/%s': %s.\n", argv0, path, de->d_name, ERRSTR);

		struct dbuf linebuf = DBUF_INIT;
		struct dbuf overflow = DBUF_INIT;

		dbuf_addstr(&linebuf, de->d_name);
		dbuf_addch(&linebuf, '=');

		ssize_t bytes = freadline(fd, &linebuf, &overflow, 0);
		if (bytes < 1)
			fatal("%s: Cannot read '%s': %s.\n", argv0, de->d_name, ERRSTR);

		/* Chomp spaces around the value. */
		char* entry = dbuf_str(&linebuf);
		while (--bytes && is_trim_char(entry[bytes]))
			/* empty */;

		if (entry[bytes++] == '=')
		    env_del(entry, bytes - 1);
		else
		    env_add(strndup(entry, bytes));

		dbuf_clear(&overflow);
		dbuf_clear(&linebuf);

		if (safe_close(fd) == -1)
			clog_warning("Cannot close '%s/%s: %s (ignored).", path, de->d_name, ERRSTR);
	}
	closedir(d);
}

static enum cflag_status
opt_envdir(const struct cflag *spec, const char *arg)
{
	if (!spec)
		return CFLAG_NEEDS_ARG;

	env_envdir(arg);
	return CFLAG_OK;
}

static enum cflag_status
opt_file(const struct cflag *spec, const char *arg)
{
    if (!spec)
        return CFLAG_NEEDS_ARG;

    int fd = safe_openat(AT_FDCWD, arg, O_RDONLY);
    if (fd < 0)
        fatal("%s: Cannot open '%s': %s.\n", argv0, arg, ERRSTR);

    struct dbuf linebuf = DBUF_INIT;
    struct dbuf overflow = DBUF_INIT;

    for (;;) {
        ssize_t bytes = freadline(fd, &linebuf, &overflow, 0);
        if (bytes == 0)
            break; /* EOF */

        if (bytes < 0)
            fatal("%s: Cannot read '%s': %s.\n", argv0, arg, ERRSTR);

        /* Chomp spaces around the entry. */
        char *entry = dbuf_str(&linebuf);

        while (bytes && is_trim_char(entry[0]))
            bytes--, entry++;
        while (bytes && is_trim_char(entry[bytes]))
            bytes--;

        if (bytes-- && entry[0] != '#') {
            const char *equalsign = strchr(entry, '=');
            assert(equalsign);
            if (equalsign - entry + 1 == bytes)
                env_del(entry, equalsign - entry);
            else
                env_add(strndup(entry, bytes));
        }

        dbuf_clear(&linebuf);
    }

    dbuf_clear(&overflow);
    dbuf_clear(&linebuf);
	if (safe_close(fd) == -1)
		clog_warning("Cannot close '%s: %s (ignored).\n", arg, ERRSTR);
    return CFLAG_OK;
}

static const struct cflag denv_options[] = {
	{
		.name = "inherit-env", .letter = 'I',
		.func = opt_inherit_env,
		.help = "Inherit all environment variables of the calling process.",
	},
	{
		.name = "inherit", .letter = 'i',
		.func = opt_inherit,
		.help = "Inherit an environment variable of the calling process.",
	},
    {
        .name = "environ", .letter = 'E',
        .func = opt_environ,
        .help =
            "Define an environment variable, or if no value is given, "
            "delete it. This option can be specified multiple times.",
    },
	{
		.name = "envdir", .letter = 'd',
		.func = opt_envdir,
		.help =
			"Add environment variables from the contents of files in "
			"a directory.",
	},
	{
		.name = "file", .letter = 'f',
		.func = opt_file,
		.help =
			"Add environment variables from a file in the environment.d(5)"
			" format. Note: $VARIABLE expansions are not supported.",
	},
	CFLAG_HELP,
	CFLAG_END
};

int
denv_main(int argc, char **argv)
{
	clog_init(NULL);

	/* XXX: It would be neat to have e.g. a cflag_argv0() function that did this. */
	argv0 = strrchr(argv[0], '/');
	if (argv0 == NULL)
		argv0 = argv[0];
	else
		argv0++;

	if (strcmp(argv0, "envdir") == 0) {
		if (argc < 3)
			fatal("%s: usage: %s d child\n", argv0, argv0);

		env_inherit_all();
		env_envdir(argv[1]);

		/* Adjust directly because replace_args_shift() does not touch argv[0]. */
		argv += 2;
		argc -= 2;
	} else {
		cflag_apply(denv_options, "[path] command [command-options...]", &argc, &argv);
		if (!argc)
			fatal("%s: No command specified.\n", argv0);
	}

	execvpe(argv[0], argv, env);
	fatal("%s: Cannot execute '%s': %s.\n", argv0, argv[0], ERRSTR);
}
